jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java

```diff
@@ -1731,6 +1739,109 @@ void destroyListener(EventListener listener)
+                    pattern = "";
+                    break;
+                case DEFAULT:
+                    mappingMatch = MappingMatch.DEFAULT;
+                    matchValue = "";
+                    break;
+                case EXACT:
+                    mappingMatch = MappingMatch.EXACT;
+                    matchValue = servletPath;
+                    if (matchValue.startsWith("/"))
+                        matchValue = matchValue.substring(1);
+                    break;
+                case PREFIX_GLOB:
+                    mappingMatch = MappingMatch.PATH;
+                    matchValue = servletPath;
+                    if (matchValue.startsWith("/"))
+                        matchValue = matchValue.substring(1);
+                    break;
+                case SUFFIX_GLOB:
+                    mappingMatch = MappingMatch.EXTENSION;
+                    int dot = servletPath.lastIndexOf('.');
+                    matchValue = servletPath.substring(0, dot);
+                    if (matchValue.startsWith("/"))
+                        matchValue = matchValue.substring(1);
+                    pattern = pattern.startsWith("/") ? pattern.substring(1) : pattern;
+                    break;
+                case MIDDLE_GLOB:
+                    mappingMatch = null;
+                    matchValue = "";
+                    break;
+                default:
+                    throw new IllegalStateException();
+            }
+        }
+        else
+        {
+            mappingMatch = null;
+            matchValue = "";
+        }
+
+        return new JettyHttpServletMapping(matchValue, pattern, servletName, mappingMatch);
+    }
+
+    /**
+     * Extends {@link MappedResource} for {@link ServletHolder} but can be
+     * asked for a {@link HttpServletMapping} which may have been created
+     * in advance for servlets which have EXACT mappings.
+     */
+    public static class MappedServlet extends MappedResource<ServletHolder>
+    {
+        private final PathSpec _pathSpec;
+        private final ServletHolder _servletHolder;
+        private final HttpServletMapping _httpServletMapping;
+
+        MappedServlet(PathSpec pathSpec, ServletHolder servletHolder)
+        {
+            super(pathSpec, servletHolder);
+            _pathSpec = pathSpec;
+            _servletHolder = servletHolder;
+
+            // Create the HttpServletMapping only once if possible.
+            if (pathSpec instanceof ServletPathSpec)
+            {
+                switch (pathSpec.getGroup())
+                {
+                    case EXACT:
+                    case DEFAULT:
+                    case ROOT:
+                        _httpServletMapping = getServletMapping(pathSpec, pathSpec.getDeclaration(), servletHolder.getName());
+                        break;
+                    default:
+                        _httpServletMapping = null;
+                        break;
+                }
+            }
+            else
```

Comment:
if pathSpec is null then this is a named dispatch so we can probably create a `_httpServletMapping` here as well?